aten/src/ATen/native/mkldnn/Pooling.cpp

```diff
@@ -0,0 +1,188 @@
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation,
+    bool ceil_mode) {
+  AT_ERROR(
+      "mkldnn_max_pool2d_with_indices: ATen not compiled with MKLDNN support");
+}
+
+std::tuple<Tensor&, Tensor&> mkldnn_max_pool2d_with_indices_out(
+    Tensor& output,
+    Tensor& indices,
+    const Tensor& self,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation,
+    bool ceil_mode) {
+  AT_ERROR(
+      "mkldnn_max_pool2d_with_indices_out: ATen not compiled with MKLDNN support");
+}
+
+Tensor mkldnn_avg_pool2d(
+    const Tensor& self,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    bool ceil_mode,
+    bool count_include_pad) {
+  AT_ERROR("mkldnn_avg_pool2d: ATen not compiled with MKLDNN support");
+}
+
+Tensor& mkldnn_avg_pool2d_out(
+    Tensor& output,
+    const Tensor& self,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    bool ceil_mode,
+    bool count_include_pad) {
+  AT_ERROR("mkldnn_avg_pool2d_out: ATen not compiled with MKLDNN support");
+}
+} // namespace native
+} // namespace at
+
+#else // AT_MKLDNN_EBABLED
+
+#include <ATen/native/mkldnn/MKLDNNCommon.h>
+#include <ATen/native/mkldnn/Utils.h>
+
+namespace at {
+namespace native {
+
+static Tensor ideep_pool2d(
+    const Tensor& input,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation,
+    bool ceil_mode,
+    ideep::algorithm algo) {
+  auto kernel_size_vec = expand_param_if_needed(kernel_size, 2);
+  auto stride_vec = expand_param_if_needed(stride, 2);
+  auto padding_vec = expand_param_if_needed(padding, 2);
+  auto dilation_vec = expand_param_if_needed(dilation, 2);
+
+  const ideep::tensor& x = itensor_from_mkldnn(input);
+  const std::vector<int64_t> output_sizes = pool_output_sizes(
+      input.sizes(),
+      kernel_size_vec,
+      stride_vec,
+      padding_vec,
+      dilation_vec,
+      ceil_mode);
+  ideep::tensor y;
+  ideep::pooling_forward::compute<AllocForMKLDNN>(
+      x,
+      {output_sizes.cbegin(), output_sizes.cend()},
+      y,
+      {stride_vec.cbegin(), stride_vec.cend()},
+      {kernel_size_vec.cbegin(), kernel_size_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      algo,
+      ideep::prop_kind::forward);
+
+  return new_with_itensor_mkldnn(std::move(y), input.options());
+}
+
+std::tuple<Tensor, Tensor> mkldnn_max_pool2d_with_indices(
+    const Tensor& input,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation,
+    bool ceil_mode) {
+  Tensor output = ideep_pool2d(
+      input,
+      kernel_size,
+      stride,
+      padding,
+      dilation,
+      ceil_mode,
+      ideep::algorithm::pooling_max);
+  return std::make_tuple(
+      output,
+      // TODO: ideep currently doesn't expose the max indices,
```

Comment:
umm...would it work to just throw an exception if they ask for the indices?

Reply:
There is no way to know whether indices is required or not in c++. We *could* throw in python side (where `return_indices` is an argument) but that would require exposing is_mkldnn to python side (not necessary a bad idea).
We *could* get the indices from mkldnn as well from a non-public api (@Jianhui-Li mentioned we can get it from the ideep::tensor `get_extra`)

Reply:
Yes, we can use `get_extra` of the output ideep tensor. See this:
https://github.com/intel/ideep/blob/960ee3a20d975655d3fa8e40e445898a1e3b4078/include/ideep/computations.hpp#L3882

Reply:
1. the problem here is that `indice` of mkldnn pooling is different from what PyTorch looks like, both the **dtype** and **content**.
a. mkldnn - `indice` is a local view w.r.t. kernel, the `dtype` could be either `int32_t` or `int8_t`
b. PyTorch - `indice` is a global view w.r.t. input.
See [mkldnn_issue_259](https://github.com/intel/mkl-dnn/issues/259) for detail.
So directly returning mkldnn indice is meaningless. The mkldnn indice is more like a cache used by backward during training. Well in that case, it should be handled by `workspace` in ATen.
2. at the current moment, probably the most simple way to address this is remove mkldnn path for `mkldnn_max_pool2d_with_indices_out`. Which means, only when `indice` is not in the output, mkldnn is used.
This won't make any difference for models like `ResNet` or `ResNext`.
3. in case we need to support unpooling (up-sampling) where `indice` is needed, we can convert the mkldnn `indice` to global by doing the math pointed in the link above.

Reply:
@bddppq Due to the differences between PyTorch indices tensor (using global indices) and MKL-DNN indices tensor (using local indices), we are considering to hide the MKL-DNN indices inside ideep as an "opaque" format. Users can get PyTorch indices with `to_dense()`. This may take a while and before that happens, how about we only implement `max_pool2d` without indices in MKL-DNN path as @mingfeima proposed?

Reply:
@bddppq - I think everyone would be calling max pool from nn.functional, not as the direct function in torch._C._nn. So I think we can change https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L489 to just call torch._C._nn.max_pool_2d and overload that function to go to mkldnn implementation. Would that work?

Reply:
aten max_pool2d right now doesn’t have dispatch (and is not even exposed to python), so in order to do that we will need to add derivatives for it to let codegen do the unwrap thing for it.

Reply:
If you just want unwrap - there's 'require_tensor: True' annotation you can put in native_functions.yaml

As for exposing to python, I think if you make it as 'python_module: nn' or somethingl ike this, it'd be exposed only as torch._C._nn - I think it's ok to have it but not call it from functional wrappers

Reply:
It's explicitly not exposed to python: https://github.com/pytorch/pytorch/blob/1c5073f/tools/autograd/gen_python_functions.py#L33
I don't dare to open this can of worms (actually I did at the very beginning and then ran into a test failure complaining ```derivative for max_pool2d is not implemented``` (or similar)).

Since
1. indices are only used in training
2. mkldnn/ideep will add support to expose indices in their api

Let's just return some garbage values and only support inference for now.

Reply:
Hm, I think it should have worked if we were to check requires_grad. But I don't insist.

Can you please expand this comment with more details on why garbage is fine? (pretty much what we have discussed above)

Reply:
Actually @wanchaol is working on exposing max_pool2d to python https://github.com/pytorch/pytorch/pull/19449. I will give a try to implement mkldnn for max_pool2d once that is merged